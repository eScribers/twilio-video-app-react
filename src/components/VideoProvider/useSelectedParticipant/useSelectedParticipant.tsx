import React, { createContext, useContext, useState, useEffect } from 'react';
import { Participant, Room } from 'twilio-video';
import rootStore from '../../../stores';
// import { ROOM_STATE } from '../../../utils/displayStrings';

type selectedParticipantContextType = [Participant | null, (participant: Participant) => void];

export const selectedParticipantContext = createContext<selectedParticipantContextType>(null!);

export default function useSelectedParticipant() {
  const setSelectedParticipant = participant => {
    rootStore.participantsStore.setSelectedParticipant(participant);
  };
  return [rootStore.participantsStore.selectedParticipant, setSelectedParticipant];
}

// type SelectedParticipantProviderProps = {
//   room: Room;
//   children: React.ReactNode;
// };

// export function SelectedParticipantProvider({ room, children }: SelectedParticipantProviderProps) {
//   const [selectedParticipant, _setSelectedParticipant] = useState<Participant | null>(null);
//   const setSelectedParticipant = (participant: Participant) =>
//     _setSelectedParticipant(prevParticipant => (prevParticipant === participant ? null : participant));

//   useEffect(() => {
//     const onDisconnect = () => _setSelectedParticipant(null);
//     room.on(ROOM_STATE.DISCONNECTED, onDisconnect);
//     return () => {
//       room.off(ROOM_STATE.DISCONNECTED, onDisconnect);
//     };
//   }, [room]);

//   return (
//     <selectedParticipantContext.Provider value={[selectedParticipant, setSelectedParticipant]}>
//       {children}
//     </selectedParticipantContext.Provider>
//   );
// }
