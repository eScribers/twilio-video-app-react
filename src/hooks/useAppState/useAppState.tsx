import { useContext } from 'react';
import { StateContext } from '../../state/AppStateProvider';

export function useAppState(): any {
  const context = useContext(StateContext);
  if (!context) {
    throw new Error('useAppState must be used within the AppStateProvider');
  }
  return context;
}
