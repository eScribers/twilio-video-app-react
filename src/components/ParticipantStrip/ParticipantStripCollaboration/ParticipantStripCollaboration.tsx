import React from 'react';
import { observer } from 'mobx-react-lite';
import { styled } from '@material-ui/core/styles';
import useVideoContext from '../../../hooks/useVideoContext/useVideoContext';
import useSelectedParticipant from '../../VideoProvider/useSelectedParticipant/useSelectedParticipant';
import Participant from '../../Participant/Participant';
import useIsSilenced from '../../../hooks/useIsSilenced/useIsSilenced';
// import useSortedParticipants from '../../../hooks/useSortedParticipants/useSortedParticipants';
import useDominantSpeaker from '../../../hooks/useDominantSpeaker/useDominantSpeaker';
import rootStore from '../../../stores';

const { participantStore } = rootStore;

const Container = styled('aside')(({ theme }) => ({
  padding: '0.5em',
  overflowY: 'scroll',
  [theme.breakpoints.down('xs')]: {
    overflowY: 'initial',
    overflowX: 'auto',
    padding: 0,
    display: 'flex',
  },
}));

const ScrollContainer = styled('div')(({ theme }) => ({
  [theme.breakpoints.down('xs')]: {
    display: 'flex',
  },
}));

const ParticipantStripCollaboration = observer(() => {
  const {
    room: { localParticipant },
  } = useVideoContext();
  const dominantSpeaker = useDominantSpeaker();
  const { sortedParticipants, selectedParticipant } = participantStore;

  const [isSilenced] = useIsSilenced();

  const dominantIdentity = dominantSpeaker?.identity;

  return (
    <Container>
      <ScrollContainer>
        <Participant
          participant={localParticipant}
          isSelected={selectedParticipant === localParticipant}
          onClick={() => participantStore.setSelectedParticipant(localParticipant)}
        />
        {sortedParticipants.map(participant => (
          <Participant
            key={participant.sid}
            isDominantSpeaker={participant.identity === dominantIdentity}
            participant={participant}
            isSelected={selectedParticipant === participant}
            userIsSilenced={!!isSilenced}
            onClick={() => participantStore.setSelectedParticipant(participant)}
          />
        ))}
      </ScrollContainer>
    </Container>
  );
});

export default ParticipantStripCollaboration;
