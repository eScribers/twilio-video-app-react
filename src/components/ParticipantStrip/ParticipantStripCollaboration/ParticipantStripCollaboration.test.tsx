import EventEmitter from 'events';
import React from 'react';
import ParticipantStripCollaboration from './ParticipantStripCollaboration';
import { shallow } from 'enzyme';
import useSelectedParticipant from '../../VideoProvider/useSelectedParticipant/useSelectedParticipant';
import useVideoContext from '../../../hooks/useVideoContext/useVideoContext';
import { useAppState } from '../../../hooks/useAppState/useAppState';

const mockUseAppState = useAppState as jest.Mock<any>;

mockUseAppState.mockImplementation(() => ({ activeSinkId: '' }));
jest.mock('../../../hooks/useAppState/useAppState');
jest.mock('../../../hooks/useVideoContext/useVideoContext');
jest.mock('../../VideoProvider/useSelectedParticipant/useSelectedParticipant');
const mockedVideoContext = useVideoContext as jest.Mock<any>;
const mockUseSelectedParticipant = useSelectedParticipant as jest.Mock<any>;
const mockLocalTrack = {
  kind: 'audio',
  mediaStreamTrack: {
    label: 'mock local audio track',
    getSettings: () => ({ deviceId: '234' }),
  },
  restart: jest.fn(),
  on: jest.fn(),
  off: jest.fn(),
};
const mockLocalParticipant = {};

describe('the ParticipantStrip component', () => {
  mockUseSelectedParticipant.mockImplementation(() => [null, () => {}]);

  it('should correctly render ParticipantInfo components', () => {
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, { sid: 0 }],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = mockLocalParticipant;
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom, localTracks: [mockLocalTrack] }));
    const wrapper = shallow(<ParticipantStripCollaboration />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should add the isSelected prop to the local participant when it is selected', () => {
    mockUseSelectedParticipant.mockImplementation(() => [mockLocalParticipant, () => {}]);
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, { sid: 0 }],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = mockLocalParticipant;
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom, localTracks: [mockLocalTrack] }));
    const wrapper = shallow(<ParticipantStripCollaboration />);
    expect(
      wrapper
        .find('Participant')
        .at(0)
        .prop('isSelected')
    ).toBe(true);
  });

  it('should add the isSelected prop to the first remote participant when it is selected', () => {
    const mockParticipant = { sid: 0 };
    mockUseSelectedParticipant.mockImplementation(() => [mockParticipant, () => {}]);
    const mockRoom: any = new EventEmitter();
    mockRoom.participants = new Map([
      [0, mockParticipant],
      [1, { sid: 1 }],
    ]);
    mockRoom.localParticipant = mockLocalParticipant;
    mockedVideoContext.mockImplementation(() => ({ room: mockRoom, localTracks: [mockLocalTrack] }));
    const wrapper = shallow(<ParticipantStripCollaboration />);
    expect(
      wrapper
        .find('Participant')
        .at(1)
        .prop('isSelected')
    ).toBe(true);
  });
});
