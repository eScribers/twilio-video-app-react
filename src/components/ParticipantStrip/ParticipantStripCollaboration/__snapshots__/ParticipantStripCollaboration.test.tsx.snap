// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ParticipantStrip component should correctly render ParticipantInfo components 1`] = `
<Styled(aside)>
  <Styled(div)>
    <Memo(wrappedComponent)
      isSelected={false}
      onClick={[Function]}
      participant={
        Object {
          "addListener": [Function],
          "audioTracks": Map {},
          "dataTracks": Map {},
          "emit": [Function],
          "eventNames": [Function],
          "getMaxListeners": [Function],
          "identity": "local@Reporter@1",
          "listenerCount": [Function],
          "listeners": [Function],
          "networkQualityLevel": null,
          "networkQualityStats": null,
          "off": [Function],
          "once": [Function],
          "prependListener": [Function],
          "prependOnceListener": [Function],
          "publishTrack": [Function],
          "publishTracks": [Function],
          "rawListeners": [Function],
          "removeAllListeners": [Function],
          "removeListener": [Function],
          "setMaxListeners": [Function],
          "setNetworkQualityConfiguration": [Function],
          "setParameters": [Function],
          "sid": "123",
          "signalingRegion": "todo",
          "state": "todo",
          "tracks": Map {},
          "unpublishTrack": [Function],
          "unpublishTracks": [Function],
          "videoTracks": Map {},
        }
      }
    />
    <Memo(wrappedComponent)
      isDominantSpeaker={false}
      isSelected={false}
      key="1"
      onClick={[Function]}
      participant={
        mockParticipant {
          "addListener": [Function],
          "audioTracks": Map {},
          "dataTracks": Map {},
          "emit": [Function],
          "eventNames": [Function],
          "getMaxListeners": [Function],
          "identity": "participant1",
          "listenerCount": [Function],
          "listeners": [Function],
          "networkQualityLevel": null,
          "networkQualityStats": null,
          "off": [Function],
          "once": [Function],
          "prependListener": [Function],
          "prependOnceListener": [Function],
          "rawListeners": [Function],
          "removeAllListeners": [Function],
          "removeListener": [Function],
          "setMaxListeners": [Function],
          "sid": "1",
          "state": "todo",
          "tracks": Map {},
          "videoTracks": Map {},
        }
      }
      userIsSilenced={false}
    />
    <Memo(wrappedComponent)
      isDominantSpeaker={false}
      isSelected={false}
      key="2"
      onClick={[Function]}
      participant={
        mockParticipant {
          "addListener": [Function],
          "audioTracks": Map {},
          "dataTracks": Map {},
          "emit": [Function],
          "eventNames": [Function],
          "getMaxListeners": [Function],
          "identity": "participant2",
          "listenerCount": [Function],
          "listeners": [Function],
          "networkQualityLevel": null,
          "networkQualityStats": null,
          "off": [Function],
          "once": [Function],
          "prependListener": [Function],
          "prependOnceListener": [Function],
          "rawListeners": [Function],
          "removeAllListeners": [Function],
          "removeListener": [Function],
          "setMaxListeners": [Function],
          "sid": "2",
          "state": "todo",
          "tracks": Map {},
          "videoTracks": Map {},
        }
      }
      userIsSilenced={false}
    />
  </Styled(div)>
</Styled(aside)>
`;
